_posts/2023-01-12-websockets.md

```diff
@@ -0,0 +1,171 @@
+That is in contrast with the **HTTP** client-server model, where a server cannot initiate a 
+message transfer to a client but can only respond to an **HTTP** request started by a client. 
+Further, by keeping the communication channel open, a WebSockets server allows the simultaneous sending and receiving 
+of messages among the connected clients through the server itself. 
+
+<center>
+    <img src="{{ site.url }}{{ site.baseurl }}/assets/images/websockets/1.png" width="400">
+    <br>
+    <em> A WebSocket server receives messages and notifies subscribers (Image generated by the author with E-Dell)</em>
+</center>
+
+In this work, we employed state-of-the-art, open-source, **JVM** technologies to assess the performances of a WebSocket server under progressive system overload. The focus of the study is comparing two event-loop models: **Spring Webflux** and **Ktor**.
+Nonetheless, some room for a didactical discussion will be left, explicitly showing the performance gap achieved by employing a thread-per-request Spring Boot WebSockets engine.
+
+## Methods
+Our primary goal is to think about a minimal system involving WebSockets communication so that all the side effects (e.g., latency) are negligible. 
+That is done by implementing an echo server, in which both the server and client run on a local machine.
+This system forwards to the client the message the client has just sent. Such a simple toy model allows us to apply progressive system overload by increasing the number of concurrent users exchanging messages with the echo server. 
+The run is carried out with **Apache JMeter** and is structured in the following way:
+- We consider an initial chunk of 50 users connected to the echo server. 
+- We keep increasing connections in chunks of 200 up to the **maximum overload state**, with 2000 concurrent users.
+- We keep the system in the maximum overload state for 3 minutes.
+- We start dropping out connections in chunks of 400 until the number of active connections drops to 0.
+If any communication errors happen, the existing connection will be closed, kicking out the user from the echo server.
+
+Separate runs for each of the two event-loop technologies are performed. The following quantities are of interest:
+- raw data describing the server response over time.
+- the actual number of concurrent users over time (may be different from the planned one, since if a communication error happens, the connection will be closed).
+- the percentage of requests processed with an error outcome.
+- the number of exchanged bytes between the concurrent clients and the echo server.
+- the echo-server throughput.
+
+Finally, we perform the simulation using a **5 kb** text sample to exchange between the client and server. 
+In particular, to simulate a real-life scenario, we require every user to exchange messages with the echo server according to a time-based Gaussian distribution. 
```

Comment:
Please specify more explicitly that the gaussian distribution is about a time delay in seconds. It become clear only after reading the second paragraph here.